File: Data_Structure/javascript/11286.js
const Heapq = function () {
  this.data = [null];
  this.size = 0;
};

Heapq.prototype.swap = function (a, b) {
  [this.data[a], this.data[b]] = [this.data[b], this.data[a]];
};
Heapq.prototype.heappush = function (item) {
  let childIdx = ++this.size;
  let parentIdx = ~~(childIdx / 2);
  this.data.push(item);
  while (parentIdx >= 1) {
    if (this.check(parentIdx, childIdx)) {
      this.swap(parentIdx, childIdx);
      [childIdx, parentIdx] = [parentIdx, ~~(parentIdx / 2)];
      continue;
    }
    return;
  }
};

Heapq.prototype.heappop = function () {
  if (this.size === 0) return 0;
  let parentIdx = 1;
  this.swap(parentIdx, this.size);
  while (true) {
    let childIdx = parentIdx * 2;
    if (childIdx + 1 < this.size && this.check(childIdx, childIdx + 1)) {
      childIdx += 1;
    }
    if (childIdx < this.size && this.check(parentIdx, childIdx)) {
      this.swap(parentIdx, childIdx);
      parentIdx = childIdx;
      continue;
    }
    this.size -= 1;
    return this.data.pop();
  }
};
Heapq.prototype.check = function (a, b) {
  const one = Math.abs(this.data[a]) > Math.abs(this.data[b]);
  const two = Math.abs(this.data[a]) === Math.abs(this.data[b]);
  const three = this.data[a] > this.data[b];
  return one || (two && three);
};

const input = require("fs").readFileSync("/dev/stdin").toString().split("\n").map(Number);
let result = "";
const heap = new Heapq();
for (let i = 1; i < input.length - 1; i++) {
  if (input[i] !== 0) {
    heap.heappush(input[i]);
  } else {
    result += heap.heappop() + "\n";
  }
}

console.log(result);

File: Data_Structure/python/11286.py
from heapq import heappush,heappop
import sys
input = sys.stdin.readline

n = int(input())
minus_heapq = []
plus_heapq = []
idx = 0
for i in range(n):
    num = int(input())
    m = False
    p = False
    if num>0:
        heappush(plus_heapq,num)
    elif num<0:
        heappush(minus_heapq,-num)
    else:
        if minus_heapq:
            m = heappop(minus_heapq)
        if plus_heapq:
            p = heappop(plus_heapq)
        if m and not p:
            print(-m)
            continue
        if p and not m:
            print(p)
            continue
        if not p and not m:
            print(0)
            continue
        if m<=p:
            heappush(plus_heapq,p)
            print(-m)
        else:
            heappush(minus_heapq,m)
            print(p)